fat/fat2/transaction.go

```diff
@@ -0,0 +1,168 @@
+package fat2
+
+import (
+	"fmt"
+)
+
+// Transaction represents a fat2 transaction, which can be a normal account
+// transaction or a conversion transaction depending on the Inputs and Outputs
+type Transaction struct {
+	Input    TransactionIO   `json:"input"`
+	Outputs  []TransactionIO `json:"outputs"`
+	Metadata interface{}     `json:"metadata,omitempty"`
+}
+
+// Valid performs all validation checks and returns nil if t is a valid Transaction
+func (t *Transaction) Valid() error {
+	// Perform initial validation on all inputs and outputs in isolation
+	if err := t.Input.Valid(); err != nil {
+		return fmt.Errorf("invalid input: %v", err)
+	}
+	if t.Input.Address == nil {
+		return fmt.Errorf("invalid input: no address")
+	} else if t.Input.Address.String() == "FA1zT4aFpEvcnPqPCigB3fvGu4Q4mTXY22iiuV69DqE1pNhdF2MC" {
+		return fmt.Errorf("invalid input: FA1zT4aFpEvcnPqPCigB3fvGu4Q4mTXY22iiuV69DqE1pNhdF2MC reserved as burn address")
+	}
+	if t.Input.Type == nil {
+		return fmt.Errorf("invalid input: no token type")
+	}
+	if len(t.Outputs) == 0 {
+		return fmt.Errorf("at least one output required")
+	}
+	for i, output := range t.Outputs {
+		if err := output.Valid(); err != nil {
+			return fmt.Errorf("invalid output at index %d: %v", i, err)
+		}
+	}
+
+	// TODO: Perform contextual validation on the collection of inputs and outputs
+	inputHasAmount := t.Input.Amount != nil
+	for i, output := range t.Outputs {
+		outputHasAddress := output.Address != nil
+		outputHasType := output.Type != nil
+		outputHasAmount := output.Amount != nil
+		outputIsConversion := outputHasType && *output.Type != *t.Input.Type
+
+		// Conversions must stay within a given address
+		if outputIsConversion && outputHasAddress && *t.Input.Address != *output.Address {
+			return fmt.Errorf("invalid output at index %d: conversion output address != input address", i)
+		}
+		// Either input or output amount can be implicit, but not both
+		if !inputHasAmount && !outputHasAmount {
+			return fmt.Errorf("invalid output at index %d: output amount required if input amount is nil", i)
+		}
+	}
+	return nil
+}
+
+// HasConversion returns true if this transaction contains a conversion input/output pair.
+// This function assumes that Transaction.Valid() returns nil
+func (t *Transaction) HasConversions() bool {
+	for _, output := range t.Outputs {
+		outputHasType := output.Type != nil
+		outputIsConversion := outputHasType && *output.Type != *t.Input.Type
+		if outputIsConversion {
+			return true
+		}
+	}
+	return false
+}
+
+// GetBalanceDeltas returns a map of all balance deltas that result from an execution of this transaction.
+// This function assumes that t.Valid() returns nil.
+func (t *Transaction) GetBalanceDeltas(conversionRates *map[string]float64) (AddressBalanceDeltaMap, error) {
+	deltas := make(AddressBalanceDeltaMap)
+	inputAddress := *t.Input.Address
```

Comment:
don't do this to save yourself some typing. Lets use these directly in the code so its always clear what we're doing at any point without having to refer to back up here. This is application of the "don't make things easy to do, make them easy to understand" value.

Reply:
Fixed